File: Matrixpy/Src/Iterative.py
def ensure_diagonally_dominant(m):
    """
    Utility to verify and enforce diagonal dominance.
    Iterative methods (Jacobi, GS, SOR) require diagonal dominance for guaranteed convergence.
    """
    if not m.isDiagonallyDominant():
        print("Matrix is not diagonally dominant. Reordering rows to improve convergence...")
        m.makeDiagonallyDominant()
    else:
        print("Matrix is already diagonally dominant.")

def solve_jacobi(m):
    """
    Solves Ax = B using the Jacobi method.
    In Jacobi, the next approximation for x[i] is calculated using only the values 
    from the previous iteration.
    """
    n = m.rows
    x = [0.0] * n
    x_new = [0.0] * n
    
    print(f"Starting Jacobi iterations (max: {m.maxIterations}, tol: {m.tolerance})...")
    
    for iteration in range(m.maxIterations):
        max_diff = 0.0
        for i in range(n):
            # Summation of A[i][j] * x_old[j] for j != i
            sigma = sum(m.mat[i][j] * x[j] for j in range(n) if i != j)
            
            if abs(m.mat[i][i]) < 1e-18:
                raise RuntimeError("Zero diagonal element detected in Jacobi.")
            
            # x_new[i] = (B[i] - sigma) / A[i][i]
            x_new[i] = (m.mat[i][n] - sigma) / m.mat[i][i]
            max_diff = max(max_diff, abs(x_new[i] - x[i]))
        
        x = list(x_new) # Synchronous update
        
        if max_diff < m.tolerance:
            print(f"Jacobi converged after {iteration + 1} iterations.")
            m.solution_vec = x
            return x
            
    print("Warning: Jacobi reached max iterations without full convergence.")
    m.solution_vec = x
    return x

def solve_gauss_seidel(m):
    """
    Solves Ax = B using the Gauss-Seidel method.
    Unlike Jacobi, Gauss-Seidel uses the newly calculated values of x[i] 
    as soon as they are available within the same iteration.
    """
    n = m.rows
    x = [0.0] * n
    
    print(f"Starting Gauss-Seidel iterations (max: {m.maxIterations}, tol: {m.tolerance})...")
    
    for iteration in range(m.maxIterations):
        max_diff = 0.0
        for i in range(n):
            old_xi = x[i]
            # Uses updated values for j < i and old values for j > i
            sigma = sum(m.mat[i][j] * x[j] for j in range(n) if i != j)
            
            if abs(m.mat[i][i]) < 1e-18:
                raise RuntimeError("Zero diagonal element detected in Gauss-Seidel.")
                
            x[i] = (m.mat[i][n] - sigma) / m.mat[i][i]
            max_diff = max(max_diff, abs(x[i] - old_xi))
            
        if max_diff < m.tolerance:
            print(f"Gauss-Seidel converged after {iteration + 1} iterations.")
            m.solution_vec = x
            return x
            
    print("Warning: Gauss-Seidel reached max iterations without full convergence.")
    m.solution_vec = x
    return x

def solve_sor(m, omega):
    """
    Solves Ax = B using Successive Over-Relaxation (SOR).
    SOR is a variant of Gauss-Seidel that uses a relaxation factor (omega).
    - 0 < omega < 1: Under-relaxation (for non-convergent systems).
    - 1 < omega < 2: Over-relaxation (to accelerate convergence).
    """
    n = m.rows
    x = [0.0] * n
    
    print(f"Starting SOR iterations (omega={omega}, max: {m.maxIterations})...")
    
    for iteration in range(m.maxIterations):
        max_diff = 0.0
        for i in range(n):
            old_xi = x[i]
            sigma = sum(m.mat[i][j] * x[j] for j in range(n) if i != j)
            
            if abs(m.mat[i][i]) < 1e-18:
                raise RuntimeError("Zero diagonal element detected in SOR.")
            
            # GS step
            x_gs = (m.mat[i][n] - sigma) / m.mat[i][i]
            
            # Weighted average with previous value
            x[i] = (1 - omega) * old_xi + omega * x_gs
            
            max_diff = max(max_diff, abs(x[i] - old_xi))
            
        if max_diff < m.tolerance:
            print(f"SOR converged after {iteration + 1} iterations.")
            m.solution_vec = x
            return x
            
    print("Warning: SOR reached max iterations.")
    m.solution_vec = x
    return x

File: Matrix1/Src/Iterative.cpp
#include "Iterative.hpp"
#include <cmath>
#include <iomanip>

// ============================================================
// Shared Iterative Helper
// ============================================================
void Iterative::ensureDiagonallyDominant(Matrix &system)
{
    if (!system.isDiagonallyDominant())
    {
        std::cout << "Matrix is not diagonally dominant. Reordering rows..." << std::endl;
        system.makeDiagonallyDominant();
    }
    else
    {
        std::cout << "Matrix is already diagonally dominant." << std::endl;
    }
}

void Iterative::ensureDiagonallyDominant()
{
    Iterative::ensureDiagonallyDominant(static_cast<Matrix &>(*this));
}

// ============================================================
// Implementation of Jacobi Method
// ============================================================
std::vector<long double> Jacobi::solveIterative()
{
    int n = rows;
    // We assume the augmented matrix is [A | b]
    // A has n rows and n columns. b is the (n+1)-th column.
    
    // Initial guess: all zeros
    std::vector<long double> x(n, 0.0);
    std::vector<long double> x_new(n, 0.0);

    std::cout << "Starting Jacobi iterations (max: " << maxIterations << ")..." << std::endl;

    for (int iter = 0; iter < maxIterations; iter++)
    {
        long double max_diff = 0.0;

        for (int i = 0; i < n; i++)
        {
            long double sigma = 0.0;
            for (int j = 0; j < n; j++)
            {
                if (i != j)
                {
                    // mat[i][j] is the coefficient from A
                    sigma += mat[i][j] * x[j];
                }
            }

            // Jacobi formula: x_new[i] = (b[i] - sigma) / A[i][i]
            // b[i] is at mat[i][n]
            if (std::abs(mat[i][i]) < 1e-18) {
                throw std::runtime_error("Zero diagonal element in Jacobi! Matrix might not be diagonally dominant.");
            }
            
            x_new[i] = (mat[i][n] - sigma) / mat[i][i];

            // Check how much it changed for convergence
            max_diff = std::max(max_diff, (long double)std::abs(x_new[i] - x[i]));
        }

        x = x_new;

        // If the change is smaller than our tolerance, we are done!
        if (max_diff < tolerance)
        {
            std::cout << "Jacobi converged after " << iter + 1 << " iterations." << std::endl;
            solution_vec = x;
            return x;
        }
    }

    std::cout << "Warning: Jacobi reached max iterations without full convergence." << std::endl;
    solution_vec = x;
    return x;
}

// ============================================================
// Implementation of Gauss-Seidel Method
// ============================================================
std::vector<long double> GaussSeidel::solveIterative()
{
    int n = rows;
    
    // Initial guess: all zeros
    std::vector<long double> x(n, 0.0);

    std::cout << "Starting Gauss-Seidel iterations (max: " << maxIterations << ")..." << std::endl;

    for (int iter = 0; iter < maxIterations; iter++)
    {
        long double max_diff = 0.0;

        for (int i = 0; i < n; i++)
        {
            long double old_xi = x[i];
            long double sigma = 0.0;
            for (int j = 0; j < n; j++)
            {
                if (i != j)
                {
                    // Here we use the ALREADY UPDATED values of x for j < i
                    // and the old values for j > i. This is the GS magic!
                    sigma += mat[i][j] * x[j];
                }
            }

            if (std::abs(mat[i][i]) < 1e-18) {
                throw std::runtime_error("Zero diagonal element in GS! Matrix might not be diagonally dominant.");
            }

            x[i] = (mat[i][n] - sigma) / mat[i][i];

            // Update max difference for convergence check
            max_diff = std::max(max_diff, (long double)std::abs(x[i] - old_xi));
        }

        if (max_diff < tolerance)
        {
            std::cout << "Gauss-Seidel converged after " << iter + 1 << " iterations." << std::endl;
            solution_vec = x;
            return x;
        }
    }

    std::cout << "Warning: Gauss-Seidel reached max iterations without full convergence." << std::endl;
    solution_vec = x;
    return x;
}
